File: bun/test/v3/ai-stream.test.ts
import { describe, test, expect } from "bun:test";
import type { NdjsonEvent } from "../../src/notion/v3/ai-types.ts";
import {
  normalizePatchStream,
  processInferenceStream,
} from "../../src/notion/v3/ai.ts";

// --- Helpers ---

async function* toAsync(
  events: NdjsonEvent[],
): AsyncIterable<NdjsonEvent> {
  for (const e of events) yield e;
}

async function collect(iter: AsyncIterable<NdjsonEvent>): Promise<NdjsonEvent[]> {
  const results: NdjsonEvent[] = [];
  for await (const item of iter) results.push(item);
  return results;
}

// ===================================================================
// normalizePatchStream
// ===================================================================

describe("normalizePatchStream", () => {
  test("passes through standard agent-inference events unchanged", async () => {
    const event: NdjsonEvent = {
      type: "agent-inference",
      id: "inf-1",
      value: [{ type: "text", content: "Hello" }],
      traceId: "t1",
      startedAt: 1,
      previousAttemptValues: [],
    };
    const results = await collect(normalizePatchStream(toAsync([event])));
    expect(results).toHaveLength(1);
    expect(results[0]).toEqual(event);
  });

  test("passes through record-map events", async () => {
    const event: NdjsonEvent = { type: "record-map" };
    const results = await collect(normalizePatchStream(toAsync([event])));
    expect(results).toHaveLength(1);
    expect(results[0]).toEqual(event);
  });

  test("converts patch-start + patch sequence into agent-inference events", async () => {
    const events: NdjsonEvent[] = [
      {
        type: "patch-start",
        data: { s: [] },
        version: 1,
      },
      {
        type: "patch",
        v: [
          {
            o: "a",
            p: "/s/-",
            v: {
              type: "agent-inference",
              id: "inf-1",
              value: [{ type: "text", content: "Hi" }],
              traceId: "t1",
              startedAt: 1,
              previousAttemptValues: [],
            },
          },
        ],
      },
    ];
    const results = await collect(normalizePatchStream(toAsync(events)));
    expect(results).toHaveLength(1);
    expect(results[0]!.type).toBe("agent-inference");
  });

  test("handles text append (o: x) correctly", async () => {
    const events: NdjsonEvent[] = [
      {
        type: "patch-start",
        data: {
          s: [
            {
              type: "agent-inference",
              id: "inf-1",
              value: [{ type: "text", content: "Hel" }],
              traceId: "t1",
              startedAt: 1,
              previousAttemptValues: [],
            },
          ],
        },
        version: 1,
      },
      {
        type: "patch",
        v: [{ o: "x", p: "/s/0/value/0/content", v: "lo" }],
      },
    ];
    const results = await collect(normalizePatchStream(toAsync(events)));
    expect(results).toHaveLength(1);
    const inf = results[0] as any;
    expect(inf.value[0].content).toBe("Hello");
  });

  test("handles remove (o: r) correctly", async () => {
    const events: NdjsonEvent[] = [
      {
        type: "patch-start",
        data: {
          s: [
            {
              type: "agent-inference",
              id: "inf-1",
              value: [
                { type: "text", content: "keep" },
                { type: "thinking", content: "remove" },
              ],
              traceId: "t1",
              startedAt: 1,
              previousAttemptValues: [],
            },
          ],
        },
        version: 1,
      },
      {
        type: "patch",
        v: [{ o: "r", p: "/s/0/value/1" }],
      },
    ];
    const results = await collect(normalizePatchStream(toAsync(events)));
    expect(results).toHaveLength(1);
    const inf = results[0] as any;
    expect(inf.value).toHaveLength(1);
    expect(inf.value[0].content).toBe("keep");
  });

  test("handles add (o: a) for new slots via /s/-", async () => {
    const events: NdjsonEvent[] = [
      { type: "patch-start", data: { s: [] }, version: 1 },
      {
        type: "patch",
        v: [
          {
            o: "a",
            p: "/s/-",
            v: {
              type: "agent-inference",
              id: "inf-1",
              value: [{ type: "text", content: "new slot" }],
              traceId: "t1",
              startedAt: 1,
              previousAttemptValues: [],
            },
          },
        ],
      },
    ];
    const results = await collect(normalizePatchStream(toAsync(events)));
    expect(results).toHaveLength(1);
    const inf = results[0] as any;
    expect(inf.value[0].content).toBe("new slot");
  });
});

// ===================================================================
// processInferenceStream
// ===================================================================

describe("processInferenceStream", () => {
  test("extracts response text from agent-inference events", async () => {
    const events: NdjsonEvent[] = [
      {
        type: "agent-inference",
        id: "inf-1",
        value: [{ type: "text", content: "Hello world" }],
        traceId: "t1",
        startedAt: 1,
        previousAttemptValues: [],
      },
    ];
    const result = await processInferenceStream(toAsync(events));
    expect(result.response).toBe("Hello world");
  });

  test("extracts title from title events", async () => {
    const events: NdjsonEvent[] = [
      { type: "title", value: "My Chat Title" },
      {
        type: "agent-inference",
        id: "inf-1",
        value: [{ type: "text", content: "response" }],
        traceId: "t1",
        startedAt: 1,
        previousAttemptValues: [],
      },
    ];
    const result = await processInferenceStream(toAsync(events));
    expect(result.title).toBe("My Chat Title");
  });

  test("extracts token counts from final agent-inference event", async () => {
    const events: NdjsonEvent[] = [
      {
        type: "agent-inference",
        id: "inf-1",
        value: [{ type: "text", content: "response" }],
        traceId: "t1",
        startedAt: 1,
        previousAttemptValues: [],
        finishedAt: 2,
        inputTokens: 100,
        outputTokens: 50,
        cachedTokensRead: 10,
        model: "oatmeal-cookie",
      },
    ];
    const result = await processInferenceStream(toAsync(events));
    expect(result.tokens).toEqual({ input: 100, output: 50, cached: 10 });
    expect(result.model).toBe("oatmeal-cookie");
  });

  test("strips lang tag from response", async () => {
    const events: NdjsonEvent[] = [
      {
        type: "agent-inference",
        id: "inf-1",
        value: [{ type: "text", content: '<lang primary="en-US"/>Hello' }],
        traceId: "t1",
        startedAt: 1,
        previousAttemptValues: [],
      },
    ];
    const result = await processInferenceStream(toAsync(events));
    expect(result.response).toBe("Hello");
  });

  test("calls onStreamChunk for streaming", async () => {
    const chunks: string[] = [];
    const events: NdjsonEvent[] = [
      {
        type: "agent-inference",
        id: "inf-1",
        value: [{ type: "text", content: "Hello" }],
        traceId: "t1",
        startedAt: 1,
        previousAttemptValues: [],
      },
      {
        type: "agent-inference",
        id: "inf-1",
        value: [{ type: "text", content: "Hello world" }],
        traceId: "t1",
        startedAt: 1,
        previousAttemptValues: [],
      },
    ];
    await processInferenceStream(toAsync(events), (chunk) => chunks.push(chunk));
    expect(chunks).toEqual(["Hello", " world"]);
  });

  test("handles empty event stream", async () => {
    const result = await processInferenceStream(toAsync([]));
    expect(result.response).toBe("");
    expect(result.title).toBeUndefined();
    expect(result.tokens).toBeUndefined();
  });
});

File: internal/notion/v3/ai_stream_test.go
package v3

import (
	"encoding/json"
	"strings"
	"testing"
	"time"
)

// =============================================================================
// ProcessInferenceStream
// =============================================================================

func TestProcessInferenceStream(t *testing.T) {
	t.Run("extracts response text", func(t *testing.T) {
		events := []NdjsonEvent{ev(t, `{"type":"agent-inference","id":"i","value":[{"type":"text","content":"Hello world"}],"traceId":"t","startedAt":1,"previousAttemptValues":[]}`)}
		if got := ProcessInferenceStream(events, nil); got.Response != "Hello world" {
			t.Errorf("response = %q", got.Response)
		}
	})

	t.Run("extracts title", func(t *testing.T) {
		events := []NdjsonEvent{
			ev(t, `{"type":"title","value":"My Chat Title"}`),
			ev(t, `{"type":"agent-inference","id":"i","value":[{"type":"text","content":"response"}],"traceId":"t","startedAt":1,"previousAttemptValues":[]}`),
		}
		if got := ProcessInferenceStream(events, nil); got.Title != "My Chat Title" {
			t.Errorf("title = %q", got.Title)
		}
	})

	t.Run("extracts token counts from final event", func(t *testing.T) {
		events := []NdjsonEvent{ev(t, `{"type":"agent-inference","id":"i","value":[{"type":"text","content":"response"}],"traceId":"t","startedAt":1,"previousAttemptValues":[],"finishedAt":2,"inputTokens":100,"outputTokens":50,"cachedTokensRead":10,"model":"oatmeal-cookie"}`)}
		got := ProcessInferenceStream(events, nil)
		if got.Model != "oatmeal-cookie" || got.Tokens == nil {
			t.Fatalf("result = %#v", got)
		}
		if got.Tokens.Input != 100 || got.Tokens.Output == nil || *got.Tokens.Output != 50 || got.Tokens.Cached == nil || *got.Tokens.Cached != 10 {
			t.Errorf("tokens = %#v", got.Tokens)
		}
	})

	t.Run("strips lang tag from response", func(t *testing.T) {
		events := []NdjsonEvent{ev(t, `{"type":"agent-inference","id":"i","value":[{"type":"text","content":"<lang primary=\"en-US\"/>Hello"}],"traceId":"t","startedAt":1,"previousAttemptValues":[]}`)}
		if got := ProcessInferenceStream(events, nil); got.Response != "Hello" {
			t.Errorf("response = %q", got.Response)
		}
	})

	t.Run("calls onChunk for streaming", func(t *testing.T) {
		var chunks []string
		events := []NdjsonEvent{
			ev(t, `{"type":"agent-inference","id":"i","value":[{"type":"text","content":"Hello"}],"traceId":"t","startedAt":1,"previousAttemptValues":[]}`),
			ev(t, `{"type":"agent-inference","id":"i","value":[{"type":"text","content":"Hello world"}],"traceId":"t","startedAt":1,"previousAttemptValues":[]}`),
		}
		ProcessInferenceStream(events, func(c string) { chunks = append(chunks, c) })
		eq(t, chunks, []string{"Hello", " world"})
	})

	t.Run("empty stream", func(t *testing.T) {
		got := ProcessInferenceStream(nil, nil)
		if got.Response != "" || got.Title != "" || got.Tokens != nil {
			t.Errorf("result = %#v", got)
		}
	})
}

// =============================================================================
// buildInferenceRequest
// =============================================================================

func TestBuildInferenceRequest(t *testing.T) {
	params := RunInferenceParams{
		Message: "hi",
		Model:   "oatmeal-cookie",
		User:    AIUser{ID: "u1", Name: "Jane", Email: "jane@example.com"},
		Space:   AISpace{ID: "space-1", Name: "WS"},
	}
	seq := 0
	uuid := func() string { seq++; return "uuid-" + string(rune('0'+seq)) }

	t.Run("new thread", func(t *testing.T) {
		req, isNew := buildInferenceRequest(params, "UTC", timeFixture(), uuid)
		if !isNew {
			t.Fatal("expected new thread")
		}
		if !req.CreateThread || !req.GenerateTitle || req.AsPatchResponse || req.IsPartialTranscript {
			t.Errorf("flags = %+v", req)
		}
		if req.ThreadParentPointer == nil || req.ThreadParentPointer.ID != "space-1" {
			t.Errorf("threadParentPointer = %#v", req.ThreadParentPointer)
		}
		if len(req.Transcript) != 3 {
			t.Fatalf("transcript len = %d", len(req.Transcript))
		}
		configVal := req.Transcript[0].(map[string]any)["value"].(map[string]any)
		if configVal["model"] != "oatmeal-cookie" || configVal["modelFromUser"] != true {
			t.Errorf("config value = %#v", configVal)
		}
		// A flag from the spread block must be present.
		if configVal["useServerUndo"] != true {
			t.Errorf("missing default config flags: %#v", configVal["useServerUndo"])
		}
	})

	t.Run("reply thread", func(t *testing.T) {
		replyParams := params
		replyParams.ThreadID = "thread-9"
		req, isNew := buildInferenceRequest(replyParams, "UTC", timeFixture(), uuid)
		if isNew {
			t.Fatal("expected reply, not new")
		}
		if req.CreateThread || req.GenerateTitle || !req.AsPatchResponse || !req.IsPartialTranscript {
			t.Errorf("flags = %+v", req)
		}
		if req.ThreadParentPointer != nil {
			t.Errorf("reply should have no threadParentPointer")
		}
		if req.ThreadID != "thread-9" {
			t.Errorf("threadId = %q", req.ThreadID)
		}
	})

	t.Run("no search", func(t *testing.T) {
		p := params
		p.NoSearch = true
		req, _ := buildInferenceRequest(p, "UTC", timeFixture(), uuid)
		configVal := req.Transcript[0].(map[string]any)["value"].(map[string]any)
		if configVal["useWebSearch"] != false {
			t.Errorf("useWebSearch = %#v", configVal["useWebSearch"])
		}
		if scopes, ok := configVal["searchScopes"].([]any); !ok || len(scopes) != 0 {
			t.Errorf("searchScopes = %#v", configVal["searchScopes"])
		}
	})

	t.Run("read-only disables edit tools", func(t *testing.T) {
		p := params
		p.ReadOnly = true
		req, _ := buildInferenceRequest(p, "UTC", timeFixture(), uuid)
		configVal := req.Transcript[0].(map[string]any)["value"].(map[string]any)
		if configVal["useReadOnlyMode"] != true {
			t.Errorf("useReadOnlyMode = %#v, want true", configVal["useReadOnlyMode"])
		}
		for _, k := range editToolFlags {
			if configVal[k] != false {
				t.Errorf("read-only should disable %s, got %#v", k, configVal[k])
			}
		}
	})

	t.Run("edits allowed keeps update-page tool on", func(t *testing.T) {
		// Default (ReadOnly false) leaves the mutating tools enabled.
		req, _ := buildInferenceRequest(params, "UTC", timeFixture(), uuid)
		configVal := req.Transcript[0].(map[string]any)["value"].(map[string]any)
		if configVal["useReadOnlyMode"] != false || configVal["enableUpdatePageV2Tool"] != true {
			t.Errorf("non-read-only config = readOnly:%v updatePage:%v", configVal["useReadOnlyMode"], configVal["enableUpdatePageV2Tool"])
		}
	})
}

// timeFixture returns a fixed time for deterministic request building.
func timeFixture() time.Time { return time.UnixMilli(1700000000000).UTC() }

// ensure the request body marshals cleanly.
func TestInferenceRequestMarshals(t *testing.T) {
	req, _ := buildInferenceRequest(RunInferenceParams{Message: "hi", Space: AISpace{ID: "s"}}, "UTC", timeFixture(), func() string { return "u" })
	if _, err := json.Marshal(req); err != nil {
		t.Fatalf("marshal: %v", err)
	}
	raw, _ := json.Marshal(req)
	if !strings.Contains(string(raw), `"threadType":"workflow"`) {
		t.Errorf("body missing threadType: %s", raw)
	}
}
